LVGL embedded GUI library

rleh opened the pull request:
- [x] Add [LVGL](https://github.com/lvgl/lvgl) as submodule
- [x] LVGL debug-log integration into modm logging
- [x] Example for STM32F469-Disco board
  - [x] Cleanup
  - [x] Use touchscreen in this example
- [x] Example for Ili9341 ~and/or SSD1306 displays~

rleh committed:
[driver] Ili9241: Add drawRaw() method

rleh commented:
CI: _Error generating documentation for device hosted-linux: 'NoneType' object is not subscriptable_

salkinium commented on examples/nucleo_l452re/lvgl/main.cpp:
```diff
@@ -0,0 +1,218 @@
+	else if(level == LV_LOG_LEVEL_TRACE) {
+		MODM_LOG_DEBUG << "LVGL(" << file << ":" << line << ") " << fn_name << ": " << dsc << modm::endl;
+	}
+}
```
Could this be moved into the `:lvgl` module? This seems very useful to have in general.

rleh replied:
Yes, but I'm unsure how to implement it.
I don't want to modify the `lv_init()` code, but we could move the code into a `lv_modm_logging.cpp` file and offer a function `lv_register_modm_logging()` to the user.
Other ideas?

salkinium replied:
You could put it into the `.init` array, so it gets called like a static constructor. This only works if you can register this before LVGL initialization.

salkinium commented on ext/lvgl/lvgl.lb:
```diff
@@ -0,0 +1,84 @@
+                "Warning": "LV_LOG_LEVEL_WARN",
+                "Error": "LV_LOG_LEVEL_ERROR",
+                },
+            default="Warning"))
```
Do we want to collapse the `logging` option into the `log_level` option with a "Disabled" entry? It would require the user to choose the log_level (ie. no default log level), but it would be one less option.

salkinium commented on ext/lvgl/lvgl.lb:
```diff
@@ -0,0 +1,84 @@
+
+- https://littlevgl.com/
+- https://github.com/littlevgl/lvgl
+"""
```
This description needs some updating ;-P

salkinium commented on ext/lvgl/lv_conf.h.in:
```diff
@@ -0,0 +1,757 @@
+ * @file lv_conf.h
+ * Configuration file for v7.11.0
+ */
+
```
```suggestion
#if __has_include(<lv_conf_local.h>)
#include <lv_conf_local.h>
#endif
```

Perhaps something like this to overwrite this file by the user. I don't really want to push all the settings through lbuild that seems stupid. We've done something like this for FreeRTOS too.

rleh replied:
... but at the end of the `lv_conf.h.in` file so that the user can overwrite the defines.

rleh replied:
How does this work for FreeRTOS?

```
In file included from modm/ext/lvgl/lvgl/src/lv_misc/../lv_conf_internal.h:38,
                 from modm/ext/lvgl/lvgl/src/lv_misc/lv_log.h:16,
                 from modm/ext/lvgl/lvgl/lvgl.h:25,
                 from modm/ext/lvgl/lvgl/lv_modm_logging.cpp:11:
modm/ext/lvgl/lv_conf.h:759:12: fatal error: lv_conf_local.h: No such file or directory
  759 | #  include <lv_conf_local.h>
      |            ^~~~~~~~~~~~~~~~~
```

File `lv_conf_local.h` exists in project top level directory and `__has_include(<lv_conf_local.h>)` is commented for debugging.

salkinium replied:
Hm… I tested this with `tusb_config_local.h` for my submodule update branch. It's just a file with an `#error "foo"` and it gets found correctly.

```
In file included from modm/ext/tinyusb/tusb_config.h:17,
                 from modm/ext/tinyusb/tusb_option.h:124,
                 from modm/ext/tinyusb/common/tusb_common.h:72,
                 from modm/ext/tinyusb/tusb.h:37,
                 from main.cpp:11:
./tusb_config_local.h:2:2: error: #error "foo"
    2 | #error "foo"
      |  ^~~~~
In file included from modm/ext/tinyusb/tusb_config.h:17,
                 from modm/ext/tinyusb/tusb_option.h:124,
                 from modm/ext/tinyusb/common/tusb_common.h:72,
                 from modm/ext/tinyusb/tusb.h:37,
                 from msc_disk.c:26:
./tusb_config_local.h:2:2: error: #error "foo"
    2 | #error "foo"
      |  ^~~~~
```

I'll check how the freertos one does it.

salkinium replied:
Hm, it's also called `FreeRTOSConfigLocal.h`… #NamingThingsConsistently…

salkinium replied:
This also works for me:
```
In file included from modm/ext/freertos/inc/freertos/FreeRTOSConfig.h:150,
                 from modm/ext/freertos/inc/freertos/FreeRTOS.h:57,
                 from modm/src/modm/processing/rtos/scheduler.hpp:18,
                 from modm/src/modm/processing/rtos.hpp:22,
                 from main.cpp:15:
./FreeRTOSConfigLocal.h:1:2: error: #error "foo"
    1 | #error "foo"
      |  ^~~~~
```

salkinium replied:
I'm using SCons, perhaps the other build systems are missing an include path…

salkinium replied:
Ah… You need to do a `scons -c` first, because it doesn't detect the dependency correctly.

salkinium replied:
Nope, I get the same error for FreeRTOS+TCP, while it works fine for FreeRTOS.

```
In file included from modm/ext/freertos_plus_tcp/include/FreeRTOS_IP.h:37,
                 from modm/ext/freertos_plus_tcp/tools/tcp_utilities/tcp_dump_packets.c:43:
modm/ext/freertos_plus_tcp/include/FreeRTOSIPConfig.h:310:10: fatal error: FreeRTOSIPConfigLocal.h: No such file or directory
  310 | #include <FreeRTOSIPConfigLocal.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

rleh replied:
```
arm-none-eabi-gcc -o /home/user/modm/build/nucleo_l452re/lvgl/scons-release/modm/ext/lvgl/lvgl/src/lv_font/lv_font_montserrat_36.o -c -std=gnu11 -Wbad-function-cast -Wimplicit -Wredundant-decls -Wstrict-prototypes -fdata-sections -ffile-prefix-map=/opt/rleh/gcc-arm-none-eabi-10-2020-q4-major=. -ffile-prefix-map=/home/user/modm/examples/nucleo_l452re/lvgl=. -ffunction-sections -finline-limit=10000 -fshort-wchar -fsingle-precision-constant -funsigned-bitfields -funsigned-char -fwrapv -g3 -gdwarf-3 -W -Wall -Wdouble-promotion -Wduplicated-cond -Werror=format -Werror=maybe-uninitialized -Werror=overflow -Werror=sign-compare -Wextra -Wlogical-op -Wno-redundant-decls -Wpointer-arith -Wundef -Os -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -mthumb -Imodm/ext -Imodm/ext/cmsis/core -Imodm/ext/cmsis/device -Imodm/ext/crashcatcher/include -Imodm/ext/lvgl -Imodm/src -Imodm modm/ext/lvgl/lvgl/src/lv_font/lv_font_montserrat_36.c

[...]

/opt/rleh/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: /home/user/modm/build/nucleo_l452re/lvgl/scons-release/main.o: in function `main':
./main.cpp:197: undefined reference to `lv_font_montserrat_36'
collect2: error: ld returned 1 exit status
```

The include path `-I.` is missing for `lv_font_montserrat_36.c` (but it is there for `main.cpp`).

Instead we have `-Imodm`, which does not look correct to me.

salkinium replied:
Yup, I found the issue a missing `abspath`. The Makefile generator also had the `-I.` missing completely… CMake did not have the issue.

rleh replied:
Fixed by @salkinium in 9a0bc1b

salkinium reviewed (commented):
I'll test in hardware and check why the docs generator is broken (yet again).

salkinium commented:
I enabled DMA2D for the STM32 devices that have it. I don't *see* any visual difference in performance, but maybe it helps elsewhere.

chris-durand commented on src/modm/driver/display/ili9341_impl.hpp:
```diff
@@ -350,6 +350,22 @@ Ili9341<Interface, Reset, Backlight, BufferSize>::drawImageRaw(glcd::Point upper
+	BatchHandle h(*this);
+
+	uint16_t* buffer = (uint16_t*)data;
+	for(size_t i = 0; i < uint16_t(width*height); i++) {
```
There is an overflow bug. This is a 320x240 display driver. Thus, `width*height` can overflow a `uint16_t`.
```suggestion
	for(size_t i = 0; i < size_t(width*height); i++) {
```

chris-durand reviewed (changes requested):
Nice!

rleh commented on ext/lvgl/lvgl.lb:
```diff
@@ -73,7 +73,7 @@ def prepare(module, options):
     module.depends(":platform:heap", ":cmsis:device")
 
-    return True
+    return options[":target"].identifier.platform in ["stm32", "sam"]
```
Why? I see no reason to do that....

salkinium replied:
Oh, I forgot LVGL also works on hosted. In that case we need to only depend on `:platform:heap` and `:cmsis:device` if the target is this, otherwise the dependencies cannot be resolved.

rleh commented:
> I enabled DMA2D for the STM32 devices that have it. I don't see any visual difference in performance, but maybe it helps elsewhere.

I have not enabled this on purpose and will disable it again because the user has to take care of configuring DMA2D/PXP manually and implementing `gpu_fill_cb`/`gpu_blend_cb`. If someone does that he can also enable DMA2D/PXP manually using `lv_conf_user.h`.

DMA2D would have to become a configurable option in any case, since it could possibly be used by a display driver as well.

salkinium committed:
[build] Fix missing local include path

salkinium commented on ext/lvgl/lv_modm_logging.cpp:
```diff
@@ -0,0 +1,42 @@
+	// register modm logging callback with LVGL
+	lv_log_register_print_cb(_lv_modm_logging_callback);
+}
+#endif
```
The constructor works just fine. Example with Info log level:

```
modm LVGL example on STM32F469-Discovery board!

LVGL(modm/ext/lvgl/lvgl/src/lv_core/lv_obj.c:231) lv_init: lv_init ready
LVGL(modm/ext/lvgl/lvgl/src/lv_core/lv_obj.c:461) lv_obj_create: Object create ready
LVGL(modm/ext/lvgl/lvgl/src/lv_core/lv_obj.c:461) lv_obj_create: Object create ready
LVGL(modm/ext/lvgl/lvgl/src/lv_core/lv_obj.c:461) lv_obj_create: Object create ready
LVGL(modm/ext/lvgl/lvgl/src/lv_core/lv_obj.c:461) lv_obj_create: Object create ready
LVGL(modm/ext/lvgl/lvgl/src/lv_widgets/lv_label.c:165) lv_label_create: label created
LVGL(modm/ext/lvgl/lvgl/src/lv_core/lv_obj.c:461) lv_obj_create: Object create ready
LVGL(modm/ext/lvgl/lvgl/src/lv_widgets/lv_cont.c:121) lv_cont_create: container created
LVGL(modm/ext/lvgl/lvgl/src/lv_widgets/lv_btn.c:106) lv_btn_create: button created
LVGL(modm/ext/lvgl/lvgl/src/lv_core/lv_obj.c:461) lv_obj_create: Object create ready
LVGL(modm/ext/lvgl/lvgl/src/lv_widgets/lv_label.c:165) lv_label_create: label created
```

salkinium commented on ext/lvgl/lvgl.lb:
```diff
@@ -0,0 +1,84 @@
+
+    env.copy("lvgl/lvgl.h")
+    env.template("lv_conf.h.in")
+    env.copy("lvgl/src")
```
The copy operator will keep the input directory structure, so you can simplify this a little.

salkinium commented on ext/lvgl/lvgl.lb:
```diff
@@ -0,0 +1,84 @@
+            default="Disabled"))
+
+    if options[":target"].identifier.platform != "hosted":
+        module.depends(":platform:heap")
```
Otherwise building hosted with the `:lvgl` module enabled (like the doxypress docs generator does) will fail because the `:platform:heap` module does not exist for hosted.

salkinium commented on ext/lvgl/lvgl.lb:
```diff
@@ -0,0 +1,84 @@
+    env.copy("lvgl/src")
+
+    if env.has_module(":debug"):
+        env.copy("lv_modm_logging.cpp")
```
Otherwise compilation may fail if people don't use our debug mechanism.

salkinium commented on tools/build_script_generator/scons/resources/SConstruct.in:
```diff
@@ -29,7 +29,7 @@ env.Alias("library", libraries)
 sources = [env.UnittestRunner(target="unittest_runner.cpp", source=headers)]
 %% else
-env.Append(CPPPATH=".")
+env.Append(CPPPATH=abspath("."))
```
🙈🙈🙈🙈

salkinium commented on tools/build_script_generator/make/resources/config.mk.in:
```diff
@@ -57,6 +57,7 @@ APP_SRC := $(APP_SRC:./%=%)
 %% for repo in generated_paths
 APP_SRC := $(filter-out {{repo}}/%,$(APP_SRC))
 %% endfor
+CPPDEFINES += -I.
```
🙈🙈🙈🙈🙈🙈🙈🙈🙈🙈🙈🙈

salkinium commented on examples/stm32f469_discovery/lvgl/main.cpp:
```diff
@@ -0,0 +1,155 @@
+	MODM_LOG_ERROR << modm::endl;
+	return modm::Abandonment::DontCare;
+}
+MODM_ASSERTION_HANDLER(test_assertion_handler);
```
Isn't this also printed by modm_abandon?

rleh replied:
I removed the fault handlers from both examples to reduce complexity.

salkinium reviewed (approved):
Perfect!

salkinium commented on ext/lvgl/lvgl.lb:
```diff
@@ -0,0 +1,84 @@
+
+- https://lvgl.io/
+- https://github.com/lvgl/lvgl
+"""
```
Ah, we should mention the `lv_conf_local.h` mechanism, I'll add a fixup commit in a second.

salkinium replied:
Hm, there's a file called `lv_conf_internal.h` that contains a *lot* of defaults.

We should also implement `LV_TICK_CUSTOM_SYS_TIME_EXPR` to use `modm::Clock::now` to remove the need for a manual call to `lv_tick_inc()`.

salkinium replied:
Ok, I've implemented and tested `LV_TICK_CUSTOM_SYS_TIME_EXPR`.

salkinium replied:
The only other thing I found interesting in terms of a modm port was `LV_ATTRIBUTE_FAST_MEM` which we could set to `__attribute__((sections(".fastcode")))`. However, 15-20kB of RAM seems like a lot and would probably cause issues if set by default. So I didn't.

```
/* Prefix performance critical functions to place them into a faster memory (e.g RAM)
 * Uses 15-20 kB extra memory */
#define  LV_ATTRIBUTE_FAST_MEM
```

salkinium commented:
Ok, I've played around with the `lv_conf.h`. The `lv_conf_internal.h` defines default values for the entire config, so I've removed all definitions (and typedefs) from `lv_conf.h` that we did not change and moved the `lv_conf_local.h` *before* our config. There are two reasons for this:

1. It removes the need for `#undef` since we're not double declaring the config. This is a very nice UX improvement.
2. It makes it easier for us to upgrade LVGL in the future, since we only need to worry about the configs that we've changed, not *all* configs.

Is this config approach ok for you?

salkinium commented:
(Sorry about the fixup! mess, I'll happily squash this all together if you want).

rleh commented:
> Ok, I've played around with the `lv_conf.h`. The `lv_conf_internal.h` defines default values for the entire config, so I've removed all definitions (and typedefs) from `lv_conf.h` that we did not change and moved the `lv_conf_local.h` _before_ our config. There are two reasons for this:
> 
>     1. It removes the need for `#undef` since we're not double declaring the config. This is a very nice UX improvement.
> 
>     2. It makes it easier for us to upgrade LVGL in the future, since we only need to worry about the configs that we've changed, not _all_ configs.
> 
> 
> Is this config approach ok for you?

Very nice, I like this approach. Thanks!

I built this PR on my attempts from >1 year ago with LittlevGL v5/v6, back then `lv_conf_internal.h` didn't exist as far as I remember; instead `lv_conf.h` was checked for completeness using `lv_conf_checker.h`.

rleh commented on ext/lvgl/lv_conf.h.in:
```diff
@@ -0,0 +1,100 @@
+ * - 8:  RGB332
+ * - 16: RGB565
+ * - 32: ARGB8888
+ */
+#define LV_COLOR_DEPTH     {{ options.color_depth }}
+
+/* Enable anti-aliasing (lines, and radiuses will be smoothed) */
+#ifndef LV_ANTIALIAS
+#define LV_ANTIALIAS        0
+#endif
+
+/* Dot Per Inch: used to initialize default sizes.
+ * E.g. a button with width = LV_DPI / 2 -> half inch wide
+ * (Not so important, you can adjust it to modify default sizes and spaces)*/
+#define LV_DPI              {{ options.dpi }}     /*[px]*/
+
+/* Type of coordinates. Should be `int16_t` (or `int32_t` for extreme cases) */
+typedef int16_t lv_coord_t;
+
+/*=========================
+   Memory manager settings
+ *=========================*/
+
+/* 1: use custom malloc/free, 0: use the built-in `lv_mem_alloc` and `lv_mem_free` */
+#define LV_MEM_CUSTOM      1
+
+/*==================
+ * Feature usage
+ *==================*/
+
+/* 1: Enable GPU interface*/
+#ifndef LV_USE_GPU
+#define LV_USE_GPU              0   /*Only enables `gpu_fill_cb` and `gpu_blend_cb` in the disp. drv- */
+#endif
+
+/* 1: Enable file system (might be required for images */
+#ifndef LV_USE_FILESYSTEM
+#define LV_USE_FILESYSTEM       0
+#endif
+
```
We could also remove all this lbuild options and instead let the user configure them through `<lv_conf_local.h>`. What do you think?

**Edit:** Except for `LV_MEM_CUSTOM`, and below `LV_TICK_CUSTOM*` and `LV_USE_LOG`/`LV_LOG_LEVEL`.

salkinium replied:
Yes, I would prefer that since the lbuild options don't add any additional benefits (ie. like other modules being able to query display size). I'll add a few more fixups.

salkinium replied:
I've moved all config options out of lbuild into `lv_conf_local.h` incl. logging, since it's weird to have to regenerate modm again just for that one option. I like this solution very much, it's very clean and minimal.

salkinium committed:
[ext] Add LVGL embedded GUI library

Co-authored-by: Raphael Lehmann <raphael@rleh.de>

rleh committed:
[examples] Add LVGL example for F469-Discovery

rleh committed:
[examples] Add LVGL example for Nucleo-L452RE with Ili9341

rleh commented:
Merge?

TomSaw commented:
Got my ili9341 on monday.
I'm happy to test LVLG - Thanks @rleh 🌻🌻🌻
